File: chap2/bribe_the_prisoners.go
package main

import "fmt"

func d(a []int) {
  for _, v := range a {
    fmt.Printf("%d\t", v)
  }
  fmt.Println()
}

func release(target int, targets []int, prisoners []int) {
  fmt.Println(target);
  d(targets)
}

func solve(p int, q int, a []int) {
  prisoners := make([]int, p)
  for i := 0; i < p; i++ { prisoners[i] = 1 }
  for i, v := range a {
    b := make([]int, len(a))
    copy(b, a)
    b = append(b[:i], b[i+1:]...)
    release(v, b, prisoners)
  }
}

func main() {
  solve(8, 1, []int { 3, 4, 5 })
}

File: chap2/bribe_the_prisoners.cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>
using namespace std;

int min_num = INT_MAX;

void solve(int num, int i, vector<int> prisoners, vector<int> targets) {
  prisoners[targets[i]-1] = false;

  for (int l = targets[i]-2; l >= 0; l--) {
    if (prisoners[l] == false) break;
    num++;    
  }

  for (int r = targets[i]; r < prisoners.size(); r++) {
    if (prisoners[r] == false) break;
    num++;    
  }

  bool last = true;
  for (int j = 0; j < targets.size(); j++) {
    if (prisoners[targets[j]-1] == true) {
      solve(num, j, prisoners, targets);
      last = false;
    }
  }

  if (last)
    if (min_num > num) min_num = num;
}

int main() {
  int p, q;
  cin >> p >> q;

  vector<int> prisoners(p);
  vector<int> targets(q);
  fill(prisoners.begin(), prisoners.end(), true);

  for (int i = 0; i < q; i++)
    cin >> targets[i];

  for (int i = 0; i < q; i++)
    solve(0, i, prisoners, targets);

  cout << min_num << endl;

  return 0;
}
